### PR TITLE
Link full Terraform plan as artifact

### PR DESCRIPTION
**Pull request by gunzip**
This is a proposal to upload the full output of Terraform plan (which is truncated into the GitHub comment) and link it into the comment.

### CODE REVIEW

File: .github/workflows/infra_plan.yaml [hunk 1]
```diff
@@ -112,16 +112,26 @@ jobs:
...
 
-          OUTPUT=$(grep -Ev "Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
+          OUTPUT=$(grep -Ev "hidden-link:|Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
 
           printf "%s" "$OUTPUT" > plan_output_multiline.txt
 
-          if grep -q "::error::Terraform exited with code" plan_output.txt; then
-            echo "failed"
+          # exit 1 if the plan failed
+          if [ "$EXIT_CODE" -ne 0 ]; then
```
Comment (gunzip) [inline .github/workflows/infra_plan.yaml:122]: rely on exit status instead of text

File: .github/workflows/infra_plan.yaml [hunk 2]
```diff
@@ -112,16 +112,26 @@ jobs:
         working-directory: ${{ steps.directory.outputs.dir }}
         run: |
           terraform plan -lock-timeout=3000s -no-color 2>&1 | grep -v "hidden-link:"  | tee plan_output.txt
+          EXIT_CODE=${PIPESTATUS[0]}
 
-          OUTPUT=$(grep -Ev "Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
+          OUTPUT=$(grep -Ev "hidden-link:|Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
```
Comment (mamu0) [inline .github/workflows/infra_plan.yaml:117]: This change is not necessary as it has already been done on line 114

File: .github/workflows/infra_plan.yaml [hunk 3]
```diff
@@ -112,16 +112,26 @@ jobs:
...
+          # exit 1 if the plan failed
+          if [ "$EXIT_CODE" -ne 0 ]; then
+            echo "Terraform plan failed with exit code $EXIT_CODE"
             exit 1
           fi
 
+      # Upload full Terraform plan output as artifact
+      - name: Upload Terraform Plan Artifact
+        if: always() && github.event_name == 'pull_request'
+        uses: actions/upload-artifact@v3
```
Comment (christian-calabrese) [inline .github/workflows/infra_plan.yaml:130]: If you agree, I would add an id to the step so we can get its output in the following step
```suggestion
        if: always() && github.event_name == 'pull_request'
        id: artifact-upload-step
        uses: actions/upload-artifact@v3
```

Comment (christian-calabrese) [inline .github/workflows/infra_plan.yaml:130]:
```suggestion
        if: always() && github.event_name == 'pull_request'
        uses: actions/upload-artifact@50769540e7f4bd5e21e526ee35c689e35e0d6874 # v4.4.0
```

File: .github/workflows/infra_plan.yaml [hunk 4]
```diff
@@ -133,6 +143,10 @@ jobs:
             const fs = require('fs');
             const output = fs.readFileSync('${{ steps.directory.outputs.dir }}/plan_output_multiline.txt', 'utf8');
             const status = '${{ steps.plan.outcome }}'
+
+            // Construct the artifacts URL
+            const artifactsUrl = `${process.env.GITHUB_SERVER_URL}/${process.env.GITHUB_REPOSITORY}/actions/runs/${process.env.GITHUB_RUN_ID}/artifacts`;
```
Comment (christian-calabrese) [inline .github/workflows/infra_plan.yaml:148]: It is possible to get from the outputs of the upload-artifact step
```suggestion
            const artifactsUrl = `${{ steps.artifact-upload-step.outputs.artifact-url }}`;
```

File: .github/workflows/infra_plan.yaml [hunk 5]
```diff
@@ -112,16 +112,26 @@ jobs:
...
+            echo "Terraform plan failed with exit code $EXIT_CODE"
             exit 1
           fi
 
+      # Upload full Terraform plan output as artifact
+      - name: Upload Terraform Plan Artifact
+        if: always() && github.event_name == 'pull_request'
+        uses: actions/upload-artifact@v3
+        with:
+          name: terraform-plan-${{ steps.directory.outputs.dir }}
```
Comment (christian-calabrese) [inline .github/workflows/infra_plan.yaml:132]: The v4 of this action does not allow the upload of the same artifact multiple times
```suggestion
          name: terraform-plan-${{ steps.directory.outputs.dir }}-${{ github.run_id }}-${{ github.run_attempt }}
```

**Comment by christian-calabrese** [comment]
@gunzip I tested the updated action in [this PR](https://github.com/pagopa/io-infra/pull/1199) and it's [getting an error](https://github.com/pagopa/io-infra/actions/runs/11108782445/job/30862551969)

File: .github/workflows/infra_plan.yaml [hunk 6]
```diff
@@ -112,16 +112,27 @@ jobs:
...
             exit 1
           fi
 
+      # Upload full Terraform plan output as artifact
+      - name: Upload Terraform Plan Artifact
+        if: always() && github.event_name == 'pull_request'
+        id: artifact-upload-step
+        uses: actions/upload-artifact@v3
+        with:
+          name: terraform-plan-${{ steps.directory.outputs.dir }}-${{ github.run_id }}-${{ github.run_attempt }}
```
Comment (gunzip) [inline .github/workflows/infra_plan.yaml:133]:
```suggestion
          name: terraform-plan-${{ github.run_id }}-${{ github.run_attempt }}
```

File: .github/workflows/infra_plan.yaml [hunk 7]
```diff
@@ -133,6 +144,10 @@ jobs:
             const fs = require('fs');
             const output = fs.readFileSync('${{ steps.directory.outputs.dir }}/plan_output_multiline.txt', 'utf8');
             const status = '${{ steps.plan.outcome }}'
+
+            // Construct the artifacts URL
+            const artifactsUrl = `${{ steps.artifact-upload-step.outputs.artifact-url }}`;
```
Comment (gunzip) [inline .github/workflows/infra_plan.yaml:149]:
```suggestion
            const artifactsUrl = `${{ steps['artifact-upload-step'].outputs['artifact-url'] }}`;
```

File: .github/workflows/infra_plan.yaml [hunk 8]
```diff
@@ -112,16 +112,27 @@ jobs:
...
+          if [ "$EXIT_CODE" -ne 0 ]; then
+            echo "Terraform plan failed with exit code $EXIT_CODE"
             exit 1
           fi
 
+      # Upload full Terraform plan output as artifact
+      - name: Upload Terraform Plan Artifact
+        if: always() && github.event_name == 'pull_request'
+        id: artifact-upload-step
+        uses: actions/upload-artifact@v3
```
Comment (gunzip) [inline .github/workflows/infra_plan.yaml:131]:
```suggestion
        uses: actions/upload-artifact@v4
```

File: .github/workflows/infra_plan.yaml [hunk 9]
```diff
@@ -133,6 +144,10 @@ jobs:
             const fs = require('fs');
             const output = fs.readFileSync('${{ steps.directory.outputs.dir }}/plan_output_multiline.txt', 'utf8');
             const status = '${{ steps.plan.outcome }}'
+
+            // Construct the artifacts URL
+            const artifactsUrl = `${{ steps['artifact-upload-step'].outputs['artifact-url'] }}`;
```
Comment (gunzip) [inline .github/workflows/infra_plan.yaml:149]:
```suggestion
            const artifactsUrl = `${{ steps.artifact-upload-step.outputs.artifact-url }}`;
```

**Comment by gunzip** [comment]
I'm closing this PR since the action to upload the plan output can only create zip files which are more impractical to read than the action output directly.

**Pull request closed without merging**